text/0000-offline-node-improved-trajectories-bags.md

```diff
@@ -0,0 +1,106 @@
+# Improved handling of multiple trajectories and bags in the offline node
+
+## Summary
+[summary]: #summary
+
+The offline node currently supports multiple sequential (non-concurrent) trajectories via loading multiple bags, where each bag is a separate trajectory.
+This RFC aims to give the user greater flexibility with the offline node when dealing with multiple robots, multiple (possibly concurrent) trajectories and multiple bags.
+
+## Motivation
+[motivation]: #motivation
+
+The online node currently supports tracking multiple concurrent trajectories in real time.
+This enables the Cartographer user to do cooperative SLAM using multiple robots.
+However, support for this in the offline node is currently limited - only a single robot configuration may be used, and the trajectories cannot be concurrent.
+Since this functionality is supported in the online node, this RFC aims to expand the offline node to be able to provide the same functionality, but at increased speed. 
+
+For example, a user has bagged a swarm of robots and wants to try out cooperative SLAM with Cartographer.
+They are interested in crunching the bags as fast as possible.
+Perhaps they have a bag for each robot, perhaps they have all robots in a single bag - the offline node should have a flexible configuration interface to allow for various possibilities.
+
+The concurrent trajectories could be processed either sequentially or in parallel - it should not influence the end result in any way, since local SLAM is not influenced by other trajectories, and global SLAM should arrive at the same pose graph at the end, no matter what the order of insertion is.
+However, processing and displaying the concurrent trajectories in parallel, like the online node can do, would be more useful and increase the coolness factor.
```

Comment:
This could also be handled by a rosbag filter or a short python script. Essentially translate each `header.stamp` in the bag to a common start time for multiple bags. We could add a script into cartographer_ros to do this. No need for the offline node to support that I think.

Reply:
To clarify: that means that the offline node could just merge sort from multiple bags: always process the newest message and push it through the pipeline. It would still work exactly the same in the current use case (just relaxing that the bags must be provided in the same time order), but would automagically support playing bags with overlapping timestamps.

Reply:
> To clarify: that means that the offline node could just merge sort from multiple bags: always process the newest message and push it through the pipeline. It would still work exactly the same in the current use case (just relaxing that the bags must be provided in the same time order), but would automagically support playing bags with overlapping timestamps.

@SirVer That's right! So there is no need to remap time in the bags, right?

I agree with you on the rest (user can split one bag into multiple bags, or join multiple bags into one, so that in the end, the data obeys 1 trajectory : 1 bag format). I think you misunderstood the text in the RFC: the user should never remap time in the bags. The offline node should accept both simultaneous and sequential trajectories. 

@wohe suggested that we could process simultaneous trajectories sequentially, to which I responded that it would be better if we processed them like you described, by merge sorting them (I called this collating, was that wrong?).

Reply:
> @SirVer That's right! So there is no need to remap time in the bags, right?

Only if you want to have the bags from one robot (i.e. the deutsches museum runs) looking like they were recorded at the same time. You know, for the coolness factor.

> I called this collating, was that wrong? Better change it to merge-sort.).

I think collating is correct as well - merge sorting is the means to get collated data here. 

> You mean the oldest? :-)

yes, sure.

Will you update the RFC with these new information?